src/plan/concurrent/immix/global.rs

```diff
@@ -0,0 +1,474 @@
+use crate::plan::concurrent::concurrent_marking_work::ProcessRootSlots;
+use crate::plan::concurrent::immix::gc_work::ConcurrentImmixGCWorkContext;
+use crate::plan::concurrent::immix::gc_work::ConcurrentImmixSTWGCWorkContext;
+use crate::plan::concurrent::Pause;
+use crate::plan::global::BasePlan;
+use crate::plan::global::CommonPlan;
+use crate::plan::global::CreateGeneralPlanArgs;
+use crate::plan::global::CreateSpecificPlanArgs;
+use crate::plan::immix::mutator::ALLOCATOR_MAPPING;
+use crate::plan::AllocationSemantics;
+use crate::plan::Plan;
+use crate::plan::PlanConstraints;
+use crate::policy::immix::ImmixSpaceArgs;
+use crate::policy::immix::TRACE_KIND_DEFRAG;
+use crate::policy::immix::TRACE_KIND_FAST;
+use crate::policy::space::Space;
+use crate::scheduler::gc_work::Release;
+use crate::scheduler::gc_work::StopMutators;
+use crate::scheduler::gc_work::UnsupportedProcessEdges;
+use crate::scheduler::*;
+use crate::util::alloc::allocators::AllocatorSelector;
+use crate::util::copy::*;
+use crate::util::heap::gc_trigger::SpaceStats;
+use crate::util::heap::VMRequest;
+use crate::util::metadata::side_metadata::SideMetadataContext;
+use crate::vm::VMBinding;
+use crate::{policy::immix::ImmixSpace, util::opaque_pointer::VMWorkerThread};
+use std::sync::atomic::AtomicBool;
+
+use atomic::Atomic;
+use atomic::Ordering;
+use enum_map::EnumMap;
+
+use mmtk_macros::{HasSpaces, PlanTraceObject};
+
+#[derive(Debug, Clone, Copy, bytemuck::NoUninit, PartialEq, Eq)]
+#[repr(u8)]
+enum GCCause {
+    Unknown,
+    FullHeap,
+    InitialMark,
+    FinalMark,
+}
+
+#[derive(HasSpaces, PlanTraceObject)]
+pub struct ConcurrentImmix<VM: VMBinding> {
+    #[post_scan]
+    #[space]
+    #[copy_semantics(CopySemantics::DefaultCopy)]
+    pub immix_space: ImmixSpace<VM>,
+    #[parent]
+    pub common: CommonPlan<VM>,
+    last_gc_was_defrag: AtomicBool,
+    current_pause: Atomic<Option<Pause>>,
+    previous_pause: Atomic<Option<Pause>>,
+    gc_cause: Atomic<GCCause>,
+    concurrent_marking_active: AtomicBool,
+}
+
+/// The plan constraints for the immix plan.
+pub const CONCURRENT_IMMIX_CONSTRAINTS: PlanConstraints = PlanConstraints {
+    // If we disable moving in Immix, this is a non-moving plan.
+    moves_objects: false,
```

Comment:
Since this plan falls back to STW GC, there is in principle nothing preventing us from doing defragmentation during STW GC.  This means this plan can still move objects.

If it is easy to enable defragmentation, we should do it so that we won't blindly need 4x-7x min heap size.  But if it is difficult or buggy, we can leave this version as fully non-moving, and enable moving in the future.

Reply:
It might be relative easy as we just need to make sure concurrent GC not doing defrag. But I never tried this so it can have bugs

Reply:
I tried locally and set `PlanConstraints::move_objects` to `true`, and it works.  But when I run a benchmark like `lusearch`, it only triggers one single STW GC in the beginning of the execution and it is a defrag GC.  Then it only triggers concurrent GC and never executes STW GC again until the benchmark finishes.  If I reduce the heap size, it will thrash, triggering concurrent GC again and again and the program will never make progress.  I guess there is something wrong with the GC trigger so that it made the GC thrash.

Reply:
`lusearch` is a bad example since it's highly generational so it's quite possible that entire blocks get freed up even in a concurrent non-moving GC. Maybe try some other benchmark?

Reply:
The concurrent GC triggering is naive, it blindly counts how many pages have been allocated and once 50% of the pages get allocated, it will start doing a concurrent GC

Reply:
Defrag is allowed now for full pauses.